diff --git a/plugins/com.aptana.editor.html/META-INF/MANIFEST.MF b/plugins/com.aptana.editor.html/META-INF/MANIFEST.MF
index 993fe92a2ca9137d974708c27dc8f555fffbdef6..242c7ddcce827b800547a5bfb538fe85674ebaf8 100644
--- a/plugins/com.aptana.editor.html/META-INF/MANIFEST.MF
+++ b/plugins/com.aptana.editor.html/META-INF/MANIFEST.MF
@@ -23,7 +23,8 @@ Require-Bundle: org.eclipse.core.runtime,
  com.aptana.parsing,
  beaver,
  com.aptana.theme,
- com.aptana.preview;visibility:=reexport
+ com.aptana.preview;visibility:=reexport,
+ com.aptana.libraries
 Bundle-RequiredExecutionEnvironment: J2SE-1.5
 Bundle-ActivationPolicy: lazy
 Bundle-Localization: plugin
diff --git a/plugins/com.aptana.editor.html/plugin.properties b/plugins/com.aptana.editor.html/plugin.properties
index 64a65d8baf2a0fab158e9a4e4b6006c7cb39100c..ee8c0ffeb98f982fea502e1860f163baa1825111 100644
--- a/plugins/com.aptana.editor.html/plugin.properties
+++ b/plugins/com.aptana.editor.html/plugin.properties
@@ -4,4 +4,6 @@ pluginName = Aptana HTML Editor
 editor.html.name = HTML Editor (Aptana)
 content-type.html.name = HTML
 
-hyperlinkDetector.cssclass.name = CSS Class Hyperlinks
\ No newline at end of file
+hyperlinkDetector.cssclass.name = CSS Class Hyperlinks
+specification.tagpair.label = HTML Tag Pair Occurrences
+validator.tidy.name = Tidy HTML Validator
\ No newline at end of file
diff --git a/plugins/com.aptana.editor.html/plugin.xml b/plugins/com.aptana.editor.html/plugin.xml
index 2d9b5ee6c47fdd8b8d4a44a0a5309e13a957d213..4965e8aa614edffe3c8eea107e7cd1dfaeca0dc5 100644
--- a/plugins/com.aptana.editor.html/plugin.xml
+++ b/plugins/com.aptana.editor.html/plugin.xml
@@ -89,7 +89,7 @@
             isGoToNextNavigationTargetKey="isOccurrenceGoToNextNavigationTarget"
             isGoToPreviousNavigationTarget="true"
             isGoToPreviousNavigationTargetKey="isOccurrenceGoToPreviousNavigationTarget"
-            label="HTML Tag Pair Occurrences"
+            label="%specification.tagpair.label"
             overviewRulerPreferenceKey="htmlTagPairOccurrenceIndicationInOverviewRuler"
             overviewRulerPreferenceValue="true"
             presentationLayer="4"
@@ -111,4 +111,12 @@
             super="com.aptana.editor.common.occurrence">
       </type>
    </extension>
+   <extension
+         point="com.aptana.editor.common.validator">
+      <validator
+            class="com.aptana.editor.html.validator.HTMLTidyValidator"
+            language="text/html"
+            name="%validator.tidy.name">
+      </validator>
+   </extension>
 </plugin>
diff --git a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/HTMLTidyValidator.java b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/HTMLTidyValidator.java
index e69de29bb2d1d6434b8b29ae775ad8c2e48c5391..1de16b5b8278dec37da09931931d52a181a97c15 100644
--- a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/HTMLTidyValidator.java
+++ b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/HTMLTidyValidator.java
@@ -0,0 +1,157 @@
+/**
+ * This file Copyright (c) 2005-2010 Aptana, Inc. This program is
+ * dual-licensed under both the Aptana Public License and the GNU General
+ * Public license. You may elect to use one or the other of these licenses.
+ * 
+ * This program is distributed in the hope that it will be useful, but
+ * AS-IS and WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, TITLE, or
+ * NONINFRINGEMENT. Redistribution, except as permitted by whichever of
+ * the GPL or APL you select, is prohibited.
+ *
+ * 1. For the GPL license (GPL), you can redistribute and/or modify this
+ * program under the terms of the GNU General Public License,
+ * Version 3, as published by the Free Software Foundation.  You should
+ * have received a copy of the GNU General Public License, Version 3 along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ * 
+ * Aptana provides a special exception to allow redistribution of this file
+ * with certain other free and open source software ("FOSS") code and certain additional terms
+ * pursuant to Section 7 of the GPL. You may view the exception and these
+ * terms on the web at http://www.aptana.com/legal/gpl/.
+ * 
+ * 2. For the Aptana Public License (APL), this program and the
+ * accompanying materials are made available under the terms of the APL
+ * v1.0 which accompanies this distribution, and is available at
+ * http://www.aptana.com/legal/apl/.
+ * 
+ * You may view the GPL, Aptana's exception and additional terms, and the
+ * APL in the file titled license.html at the root of the corresponding
+ * plugin containing this source file.
+ * 
+ * Any modifications to this file must keep this entire header intact.
+ */
+package com.aptana.editor.html.validator;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.io.StringReader;
+import java.net.URI;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.w3c.tidy.Tidy;
+
+import com.aptana.core.util.StringUtil;
+import com.aptana.editor.common.validator.IValidationManager;
+import com.aptana.editor.common.validator.IValidator;
+import com.aptana.editor.html.Activator;
+
+public class HTMLTidyValidator implements IValidator
+{
+
+	private static final Pattern PATTERN = Pattern
+			.compile("\\s*line\\s+(\\d+)\\s*column\\s+(\\d+)\\s*-\\s*(Warning|Error):\\s*(.+)$"); //$NON-NLS-1$
+
+	public void parse(String source, URI path, IValidationManager manager)
+	{
+		String report = parseWithTidy(source);
+		if (StringUtil.isEmpty(report))
+		{
+			return;
+		}
+
+		String uri = path.toString();
+		BufferedReader reader = null;
+		try
+		{
+			reader = new BufferedReader(new StringReader(report));
+			String line;
+			while ((line = reader.readLine()) != null)
+			{
+				if (line.startsWith("line")) //$NON-NLS-1$
+				{
+					parseTidyOutput(line, uri, manager);
+				}
+			}
+		}
+		catch (Exception e)
+		{
+			Activator.logError(Messages.HTMLTidyValidator_ERR_ParseErrors, e);
+		}
+		finally
+		{
+			if (reader != null)
+			{
+				try
+				{
+					reader.close();
+				}
+				catch (IOException e)
+				{
+					// ignores
+				}
+			}
+		}
+	}
+
+	private static String parseWithTidy(String source)
+	{
+		Tidy tidy = new Tidy();
+		ByteArrayOutputStream bout = new ByteArrayOutputStream();
+		PrintWriter out = new PrintWriter(bout);
+		tidy.setErrout(out);
+		try
+		{
+			tidy.parse(new ByteArrayInputStream(source.getBytes()), null);
+		}
+		catch (Exception e)
+		{
+			Activator.logError(Messages.HTMLTidyValidator_ERR_Tidy, e);
+		}
+		out.flush();
+
+		return bout.toString();
+	}
+
+	private static void parseTidyOutput(String report, String path, IValidationManager manager)
+	{
+		Matcher matcher = PATTERN.matcher(report);
+
+		while (matcher.find())
+		{
+			int lineNumber = Integer.parseInt(matcher.group(1));
+			int column = Integer.parseInt(matcher.group(2));
+			String type = matcher.group(3);
+			String message = patchMessage(matcher.group(4));
+
+			if (message != null)
+			{
+				if (type.startsWith("Error")) //$NON-NLS-1$
+				{
+					manager.addError(message, lineNumber, column, 0, path);
+				}
+				else if (type.startsWith("Warning")) //$NON-NLS-1$
+				{
+					manager.addWarning(message, lineNumber, column, 0, path);
+				}
+			}
+		}
+	}
+
+	private static String patchMessage(String message)
+	{
+		if (message == null)
+		{
+			return null;
+		}
+		message = message.replaceFirst("discarding", "should discard"); //$NON-NLS-1$ //$NON-NLS-2$
+		message = message.replaceFirst("inserting", "should insert"); //$NON-NLS-1$ //$NON-NLS-2$
+		message = message.replaceFirst("trimming", "should trim"); //$NON-NLS-1$ //$NON-NLS-2$
+		return message;
+	}
+}
diff --git a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/Messages.java b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/Messages.java
index e69de29bb2d1d6434b8b29ae775ad8c2e48c5391..8edb1f9d7f359ca0a2bd517bba25860bf06bbfcd 100644
--- a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/Messages.java
+++ b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/Messages.java
@@ -0,0 +1,56 @@
+/**
+ * This file Copyright (c) 2005-2010 Aptana, Inc. This program is
+ * dual-licensed under both the Aptana Public License and the GNU General
+ * Public license. You may elect to use one or the other of these licenses.
+ * 
+ * This program is distributed in the hope that it will be useful, but
+ * AS-IS and WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE, TITLE, or
+ * NONINFRINGEMENT. Redistribution, except as permitted by whichever of
+ * the GPL or APL you select, is prohibited.
+ *
+ * 1. For the GPL license (GPL), you can redistribute and/or modify this
+ * program under the terms of the GNU General Public License,
+ * Version 3, as published by the Free Software Foundation.  You should
+ * have received a copy of the GNU General Public License, Version 3 along
+ * with this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ * 
+ * Aptana provides a special exception to allow redistribution of this file
+ * with certain other free and open source software ("FOSS") code and certain additional terms
+ * pursuant to Section 7 of the GPL. You may view the exception and these
+ * terms on the web at http://www.aptana.com/legal/gpl/.
+ * 
+ * 2. For the Aptana Public License (APL), this program and the
+ * accompanying materials are made available under the terms of the APL
+ * v1.0 which accompanies this distribution, and is available at
+ * http://www.aptana.com/legal/apl/.
+ * 
+ * You may view the GPL, Aptana's exception and additional terms, and the
+ * APL in the file titled license.html at the root of the corresponding
+ * plugin containing this source file.
+ * 
+ * Any modifications to this file must keep this entire header intact.
+ */
+package com.aptana.editor.html.validator;
+
+import org.eclipse.osgi.util.NLS;
+
+public class Messages extends NLS
+{
+
+	private static final String BUNDLE_NAME = "com.aptana.editor.html.validator.messages"; //$NON-NLS-1$
+
+	public static String HTMLTidyValidator_ERR_ParseErrors;
+	public static String HTMLTidyValidator_ERR_Tidy;
+
+	static
+	{
+		// initialize resource bundle
+		NLS.initializeMessages(BUNDLE_NAME, Messages.class);
+	}
+
+	private Messages()
+	{
+	}
+}
diff --git a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/messages.properties b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/messages.properties
index e69de29bb2d1d6434b8b29ae775ad8c2e48c5391..a1c93afc605135c60be70f141942ebbc2f8611f3 100644
--- a/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/messages.properties
+++ b/plugins/com.aptana.editor.html/src/com/aptana/editor/html/validator/messages.properties
@@ -0,0 +1,2 @@
+HTMLTidyValidator_ERR_ParseErrors=Error parsing the HTML errors
+HTMLTidyValidator_ERR_Tidy=Error when running the source through tidy
diff --git a/plugins/com.aptana.libraries/.classpath b/plugins/com.aptana.libraries/.classpath
index 1366ffdb88cf70e7ed501b9464ca55a1cfe32484..5c297189f9b7db231aaeee3d887d32918745cd76 100644
--- a/plugins/com.aptana.libraries/.classpath
+++ b/plugins/com.aptana.libraries/.classpath
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
+	<classpathentry exported="true" kind="lib" path="Tidy.jar"/>
 	<classpathentry exported="true" kind="lib" path="velocity.jar"/>
 	<classpathentry exported="true" kind="lib" path="css-validator.jar"/>
 	<classpathentry exported="true" kind="lib" path="derby.jar"/>
diff --git a/plugins/com.aptana.libraries/META-INF/MANIFEST.MF b/plugins/com.aptana.libraries/META-INF/MANIFEST.MF
index 4cb9e9b26e6feadf0a6128daa628559cbdcbac4e..9349da2ec20f116cffe6b7b140e26892beff7c9e 100644
--- a/plugins/com.aptana.libraries/META-INF/MANIFEST.MF
+++ b/plugins/com.aptana.libraries/META-INF/MANIFEST.MF
@@ -8,7 +8,8 @@ Bundle-ClassPath: .,
  derby.jar,
  uuid.jar,
  css-validator.jar,
- velocity.jar
+ velocity.jar,
+ Tidy.jar
 Export-Package: com.eaio.uuid,
  org.apache.commons.lang,
  org.apache.derby.authentication,
@@ -133,6 +134,9 @@ Export-Package: com.eaio.uuid,
  org.w3c.css.util,
  org.w3c.css.util.xml,
  org.w3c.css.values,
- org.w3c.www.mime
+ org.w3c.dom,
+ org.w3c.tidy,
+ org.w3c.www.mime,
+ org.xml.sax
 Eclipse-BuddyPolicy: registered
 Bundle-RequiredExecutionEnvironment: J2SE-1.5
diff --git a/plugins/com.aptana.libraries/Tidy.jar b/plugins/com.aptana.libraries/Tidy.jar
index e69de29bb2d1d6434b8b29ae775ad8c2e48c5391..af4eed0506b53f17a4d22e4f1630ee03cb7991e5 100644
Binary files a/plugins/com.aptana.libraries/Tidy.jar and b/plugins/com.aptana.libraries/Tidy.jar differ
diff --git a/plugins/com.aptana.libraries/build.properties b/plugins/com.aptana.libraries/build.properties
index f147b3fd6eb23294e1ff6c1556c6b9370f2b3691..57213e14e2e3108bda3d1f53eb76ba34aa283562 100644
--- a/plugins/com.aptana.libraries/build.properties
+++ b/plugins/com.aptana.libraries/build.properties
@@ -6,4 +6,5 @@ bin.includes = META-INF/,\
                derby-LICENSE,\
                css-validator.jar,\
                velocity.jar,\
-               velocity-LICENSE
+               velocity-LICENSE,\
+               Tidy.jar
