Index: src/org/jruby/parser/DefaultRubyParser.java
===================================================================
--- src/org/jruby/parser/DefaultRubyParser.java	(revision 7993)
+++ src/org/jruby/parser/DefaultRubyParser.java	(working copy)
@@ -102,7 +102,6 @@
 import org.jruby.ast.SClassNode;
 import org.jruby.ast.SValueNode;
 import org.jruby.ast.SelfNode;
-import org.jruby.ast.SplatNode;
 import org.jruby.ast.StarNode;
 import org.jruby.ast.StrNode;
 import org.jruby.ast.SymbolNode;
@@ -125,11 +124,11 @@
 import org.jruby.lexer.yacc.ISourcePositionHolder;
 import org.jruby.lexer.yacc.LexerSource;
 import org.jruby.lexer.yacc.RubyYaccLexer;
-import org.jruby.lexer.yacc.RubyYaccLexer.LexState;
 import org.jruby.lexer.yacc.StrTerm;
 import org.jruby.lexer.yacc.SyntaxException;
+import org.jruby.lexer.yacc.Token;
+import org.jruby.lexer.yacc.RubyYaccLexer.LexState;
 import org.jruby.lexer.yacc.SyntaxException.PID;
-import org.jruby.lexer.yacc.Token;
 import org.jruby.runtime.Visibility;
 import org.jruby.util.ByteList;
 
@@ -134,12 +133,16 @@
 import org.jruby.util.ByteList;
 
 public class DefaultRubyParser {
-    private ParserSupport support;
-    private RubyYaccLexer lexer;
-    private IRubyWarnings warnings;
+    protected ParserSupport support;
+    protected RubyYaccLexer lexer;
+    protected IRubyWarnings warnings;
 
     public DefaultRubyParser() {
-        support = new ParserSupport();
+        this(new ParserSupport());
+    }
+    
+    public DefaultRubyParser(ParserSupport support) {
+        this.support = support;
         lexer = new RubyYaccLexer();
         lexer.setParserSupport(support);
     }
@@ -3328,6 +3331,7 @@
         
         lexer.reset();
         lexer.setSource(source);
+        lexer.setParserSupport(support);
         try {
 	    //yyparse(lexer, new jay.yydebug.yyAnim("JRuby", 9));
 	    //yyparse(lexer, new jay.yydebug.yyDebugAdapter());
