Index: C:/Users/Chris Williams/33-workspace/jruby_1.1_rc2/src/org/jruby/lexer/yacc/LexerSource.java
===================================================================
--- C:/Users/Chris Williams/33-workspace/jruby_1.1_rc2/src/org/jruby/lexer/yacc/LexerSource.java	(revision 6005)
+++ C:/Users/Chris Williams/33-workspace/jruby_1.1_rc2/src/org/jruby/lexer/yacc/LexerSource.java	(working copy)
@@ -33,6 +33,7 @@
 
 import java.io.IOException;
 import java.io.InputStream;
+import java.io.Reader;
 import java.util.List;
 
 import org.jruby.parser.ParserConfiguration;
@@ -147,6 +148,12 @@
         return new InputStreamLexerSource(name, content, list, configuration.getLineNumber(), 
                 configuration.hasExtraPositionInformation());
     }
+    
+    public static LexerSource getSource(String name, Reader content, List<String> list,
+            ParserConfiguration configuration) {
+        return new ReaderLexerSource(name, content, list, configuration.getLineNumber(), 
+                configuration.hasExtraPositionInformation());
+    }
 
     protected void captureFeature(int c) {
         // Ruby's OMG capture all source in a Hash feature
